In [1]:
import warnings
warnings.filterwarnings('ignore')

# importing modules and packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import pandas as pd
import tensorflow as tf
from pathlib import Path
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier

In [2]:
covid_df = pd.read_csv(Path('Covid_Data.csv'))
covid_df.head()

USMER  MEDICAL_UNIT  SEX  PATIENT_TYPE   DATE_DIED  INTUBED  PNEUMONIA  \
0      2             1    1             1  03/05/2020       97          1   
1      2             1    2             1  03/06/2020       97          1   
2      2             1    2             2  09/06/2020        1          2   
3      2             1    1             1  12/06/2020       97          2   
4      2             1    2             1  21/06/2020       97          2   

   AGE  PREGNANT  DIABETES  ...  ASTHMA  INMSUPR  HIPERTENSION  OTHER_DISEASE  \
0   65         2         2  ...       2        2             1              2   
1   72        97         2  ...       2        2             1              2   
2   55        97         1  ...       2        2             2              2   
3   53         2         2  ...       2        2             2              2   
4   68        97         1  ...       2        2             1              2   

   CARDIOVASCULAR  OBESITY  RENAL_CHRONIC  TOBACCO  CLASIFFICATION_FINAL  ICU  
0               2        2              2        2                     3   97  
1               2        1              1        2                     5   97  
2               2        2              2        2                     3    2  
3               2        2              2        2                     7   97  
4               2        2              2        2                     3   97  

[5 rows x 21 columns]

In [3]:
# Applying the condition-Outcome column
covid_df['DATE_DIED'] = ['0' if DATE_DIED == '9999-99-99' else '1' for DATE_DIED in covid_df['DATE_DIED']]
covid_df.head(30)

USMER  MEDICAL_UNIT  SEX  PATIENT_TYPE DATE_DIED  INTUBED  PNEUMONIA  AGE  \
0       2             1    1             1         1       97          1   65   
1       2             1    2             1         1       97          1   72   
2       2             1    2             2         1        1          2   55   
3       2             1    1             1         1       97          2   53   
4       2             1    2             1         1       97          2   68   
5       2             1    1             2         0        2          1   40   
6       2             1    1             1         0       97          2   64   
7       2             1    1             1         0       97          1   64   
8       2             1    1             2         0        2          2   37   
9       2             1    1             2         0        2          2   25   
10      2             1    1             1         0       97          2   38   
11      2             1    2             2         0        2          2   24   
12      2             1    2             2         0        2          2   30   
13      2             1    2             1         0       97          2   55   
14      2             1    1             1         0       97          2   48   
15      2             1    1             1         0       97          2   23   
16      2             1    1             2         0        2          1   80   
17      2             1    2             1         0       97          2   61   
18      2             1    2             1         0       97          2   54   
19      2             1    1             1         0       97          2   64   
20      2             1    2             2         0        2          1   59   
21      2             1    2             1         0       97          2   30   
22      2             1    2             1         0       97          2   45   
23      2             1    1             1         0       97          2   26   
24      2             1    1             1         0       97          2   38   
25      2             1    2             1         0       97          2   24   
26      2             1    2             1         0       97          2   32   
27      2             1    2             1         0       97          2   49   
28      2             1    2             1         0       97          2   39   
29      2             1    2             1         0       97          2   27   

    PREGNANT  DIABETES  ...  ASTHMA  INMSUPR  HIPERTENSION  OTHER_DISEASE  \
0          2         2  ...       2        2             1              2   
1         97         2  ...       2        2             1              2   
2         97         1  ...       2        2             2              2   
3          2         2  ...       2        2             2              2   
4         97         1  ...       2        2             1              2   
5          2         2  ...       2        2             2              2   
6          2         2  ...       2        2             2              2   
7          2         1  ...       2        1             1              2   
8          2         1  ...       2        2             1              2   
9          2         2  ...       2        2             2              2   
10         2         2  ...       2        2             2              2   
11        97         2  ...       2        2             2              2   
12        97         2  ...       2        2             2              2   
13        97         2  ...       2        2             2              2   
14         2         1  ...       2        2             2              2   
15         2         2  ...       2        2             2              2   
16         2         2  ...       2        2             1              2   
17        97         2  ...       2        2             2              2   
18        97         2  ...  

In [4]:
# Generate our categorical variable list
covid_df.dtypes[covid_df.dtypes == "object"].index.tolist()

['DATE_DIED']

In [5]:
# Drop NA values(values>2)and column with too much missing values
covid_df = covid_df.drop(covid_df[covid_df["PREGNANT"] >= 2].index)

In [6]:
#covid_df = covid_df.drop(["PREGNANT","INTUBED","ICU"], axis = 1)
covid_df = covid_df.drop(["INTUBED","ICU"], axis = 1)
covid_df.head()

USMER  MEDICAL_UNIT  SEX  PATIENT_TYPE DATE_DIED  PNEUMONIA  AGE  \
969       2             3    1             2         1          1   23   
1735      1             3    1             2         1          1   30   
1787      2             3    1             2         1          1   33   
1856      1             3    1             1         0          2   57   
1903      2             3    1             2         0          1   25   

      PREGNANT  DIABETES  COPD  ASTHMA  INMSUPR  HIPERTENSION  OTHER_DISEASE  \
969          1         2     2       2        2             2              2   
1735         1         2     2       2        2             2              2   
1787         1         2     2       2        2             2              1   
1856         1         2     2       2        2             1              2   
1903         1         2     2       2        2             2              2   

      CARDIOVASCULAR  OBESITY  RENAL_CHRONIC  TOBACCO  CLASIFFICATION_FINAL  
969                2        1              2        2                     3  
1735               2        2              2        2                     3  
1787               2        2              2        2                     3  
1856               2        1              2        2                     1  
1903               2        2              2        2                     1

In [7]:
# Column values
covid_df.columns.values 

array(['USMER', 'MEDICAL_UNIT', 'SEX', 'PATIENT_TYPE', 'DATE_DIED',
       'PNEUMONIA', 'AGE', 'PREGNANT', 'DIABETES', 'COPD', 'ASTHMA',
       'INMSUPR', 'HIPERTENSION', 'OTHER_DISEASE', 'CARDIOVASCULAR',
       'OBESITY', 'RENAL_CHRONIC', 'TOBACCO', 'CLASIFFICATION_FINAL'],
      dtype=object)

In [8]:
# Remove COVID death/fatality outcome target from features data
y = covid_df.DATE_DIED
X = covid_df.drop(columns="DATE_DIED")

In [9]:
X.shape

(8131, 18)

In [10]:
# Split training/test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, stratify=y)

In [11]:
# Create a StandardScaler instance
scaler = preprocessing.StandardScaler().fit(X_train)
scaler

StandardScaler()

In [12]:
scaler.mean_

array([ 1.6305346 ,  8.97310594,  1.        ,  1.24089866,  4.9180059 ,
       27.99737619,  1.        ,  2.12873073,  2.0916694 ,  2.08019023,
        2.1457855 ,  2.06559528,  2.19744178,  2.10511643,  2.03132174,
        2.10659233,  2.08707773,  5.48950476])

In [13]:
# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Logistic Regression

In [14]:
# Define the logistic regression model
log_classifier = LogisticRegression(solver="lbfgs",max_iter=1000)

#### Note:
* The solver parameter is set to 'lbfgs', which is an algorithm for learning and optimization. The particular solver isn't very important in this example, but note that a number of optimizers exist.
* The max_iter parameter will be set to 1000 iterations, which will give the model sufficient opportunity to converge on effective weights

In [15]:
# Train the model
log_classifier.fit(X_train,y_train)

# Evaluate the model
y_pred = log_classifier.predict(X_test)
print(f" Logistic regression model accuracy: {accuracy_score(y_test,y_pred):.3f}")

 Logistic regression model accuracy: 0.987


#### Interpretation:
* Logistic model able to predict correctly whether or not a person with the  underline diseases 98.7% of the time will have a high risk of developing sever COVID which leads ultimately to fatality.

## Random Forest 

In [16]:
# Create a random forest classifier.
rf_model = RandomForestClassifier(n_estimators=128, random_state=78)

# Fitting the model
rf_model = rf_model.fit(X_train_scaled, y_train)

In [17]:
# Evaluate the model
y_pred = rf_model.predict(X_test_scaled)
print(f" Random forest predictive accuracy: {accuracy_score(y_test,y_pred):.3f}")

 Random forest predictive accuracy: 0.985


#### Interpretation:
* Random forest model  able to predict correctly whether or not a person with the mentioned underline diseases 98.5% of the time will have a high risk of developing sever COVID which leads ultimately to fatality.

## Gradient Boosting Classifier

In [18]:
# Choose a learning rate and create classifier
classifier = GradientBoostingClassifier(n_estimators=20,
                                        learning_rate=0.5,
                                        max_features=5,
                                        max_depth=3,
                                        random_state=0)

In [19]:
# Fit the model
classifier.fit(X_train_scaled, y_train)

GradientBoostingClassifier(learning_rate=0.5, max_features=5, n_estimators=20,
                           random_state=0)

In [20]:
# Evaluate the model
y_pred = classifier.predict(X_test_scaled)
print(f" SVM model accuracy: {accuracy_score(y_test,y_pred):.3f}")

 SVM model accuracy: 0.987


#### Interpretation:
* Gradient Boosting Classifier model able to predict correctly whether or not a patient with the mentioned underline diseases, 98.7% of the time will have a high risk of developing sever COVID which leads ultimately to fatality.

## Support Vector Machine

In [21]:
# Create the SVM model
svm = SVC(kernel='linear')

In [22]:
# Train the model
svm.fit(X_train, y_train)

SVC(kernel='linear')

In [23]:
# Evaluate the model
y_pred = svm.predict(X_test_scaled)
print(f" SVM model accuracy: {accuracy_score(y_test,y_pred):.3f}")

 SVM model accuracy: 0.989


#### Interpretation:
* Looking at the output of our SVM model, the model was able to correctly predict  whether or not a person with the underline diseases 98.9% of the time will have a high risk of developing sever COVID which leads ultimately to fatality.

## Deep Learning

In [24]:
# Define the model - deep neural net
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 = 18
hidden_nodes_layer2 = 10

model = tf.keras.models.Sequential()

# First hidden layer
first_layer = tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
model.add(first_layer)


# Second hidden layer
second_layer = tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu")
model.add(second_layer)


# Output layer
output_layer = tf.keras.layers.Dense(units=1, activation="sigmoid")
model.add(output_layer)

# Compile the Sequential model together and customize metrics
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [25]:
# Train the model
fit_model = model.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100


UnimplementedError: Graph execution error:

Detected at node 'binary_crossentropy/Cast' defined at (most recent call last):
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\runpy.py", line 193, in _run_module_as_main
      "__main__", mod_spec)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\runpy.py", line 85, in _run_code
      exec(code, run_globals)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel_launcher.py", line 17, in <module>
      app.launch_new_instance()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
      app.start()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\kernelapp.py", line 712, in start
      self.io_loop.start()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\tornado\platform\asyncio.py", line 215, in start
      self.asyncio_loop.run_forever()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\asyncio\base_events.py", line 541, in run_forever
      self._run_once()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\asyncio\base_events.py", line 1786, in _run_once
      handle._run()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\asyncio\events.py", line 88, in _run
      self._context.run(self._callback, *self._args)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\kernelbase.py", line 510, in dispatch_queue
      await self.process_one()
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\kernelbase.py", line 499, in process_one
      await dispatch(*args)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\kernelbase.py", line 406, in dispatch_shell
      await result
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\kernelbase.py", line 730, in execute_request
      reply_content = await reply_content
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\ipkernel.py", line 390, in do_execute
      res = shell.run_cell(code, store_history=store_history, silent=silent)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\ipykernel\zmqshell.py", line 528, in run_cell
      return super().run_cell(*args, **kwargs)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\IPython\core\interactiveshell.py", line 2915, in run_cell
      raw_cell, store_history, silent, shell_futures)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\IPython\core\interactiveshell.py", line 2960, in _run_cell
      return runner(coro)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\IPython\core\async_helpers.py", line 78, in _pseudo_sync_runner
      coro.send(None)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\IPython\core\interactiveshell.py", line 3186, in run_cell_async
      interactivity=interactivity, compiler=compiler, result=result)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\IPython\core\interactiveshell.py", line 3377, in run_ast_nodes
      if (await self.run_code(code, result,  async_=asy)):
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\IPython\core\interactiveshell.py", line 3457, in run_code
      exec(code_obj, self.user_global_ns, self.user_ns)
    File "C:\Users\abram\AppData\Local\Temp\ipykernel_7424\3344224766.py", line 2, in <module>
      fit_model = model.fit(X_train_scaled, y_train, epochs=100)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\utils\traceback_utils.py", line 65, in error_handler
      return fn(*args, **kwargs)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\training.py", line 1650, in fit
      tmp_logs = self.train_function(iterator)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\training.py", line 1249, in train_function
      return step_function(self, iterator)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\training.py", line 1233, in step_function
      outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\training.py", line 1222, in run_step
      outputs = model.train_step(data)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\training.py", line 1024, in train_step
      loss = self.compute_loss(x, y, y_pred, sample_weight)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\training.py", line 1083, in compute_loss
      y, y_pred, sample_weight, regularization_losses=self.losses
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\engine\compile_utils.py", line 265, in __call__
      loss_value = loss_obj(y_t, y_p, sample_weight=sw)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\losses.py", line 152, in __call__
      losses = call_fn(y_true, y_pred)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\losses.py", line 284, in call
      return ag_fn(y_true, y_pred, **self._fn_kwargs)
    File "C:\Users\abram\anaconda3\envs\mlenv\lib\site-packages\keras\losses.py", line 2165, in binary_crossentropy
      y_true = tf.cast(y_true, y_pred.dtype)
Node: 'binary_crossentropy/Cast'
Cast string to float is not supported
	 [[{{node binary_crossentropy/Cast}}]] [Op:__inference_train_function_1086]

In [200]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")